pallets/stablewap/src/lib.rs

```diff
@@ -0,0 +1,696 @@
+		AssetNotRegistered,
+
+		/// Invalid asset amount provided. Amount must be greater than zero.
+		InvalidAssetAmount,
+
+		/// Balance of an asset is nto sufficient to perform a trade.
+		InsufficientBalance,
+
+		/// Balance of an share asset is nto sufficient to withdraw liquidity.
+		InsufficientShares,
+
+		/// Liquidity has not reached the required minimum.
+		InsufficientLiquidity,
+
+		/// Insufficient liquidity left in the pool after withdrawal.
+		InsufficientLiquidityRemaining,
+
+		/// Amount is less than min trading limit.
+		InsufficientTradingAmount,
+
+		/// Minimum limit has not been reached during trade.
+		BuyLimitNotReached,
+
+		/// Maximum limit has been exceeded during trade.
+		SellLimitExceeded,
+
+		/// Initial liquidity of asset must be > 0.
+		InvalidInitialLiquidity,
+
+		/// Account balance is too low.
+		BalanceTooLow,
+
+		/// Amplification is outside specified range.
+		InvalidAmplification,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Create a 2-asset pool with initial liquidity for both assets.
+		///
+		/// Both assets must be correctly registered in `T::AssetRegistry`
+		///
+		/// Initial liquidity must be > 0.
+		///
+		/// Origin is given corresponding amount of shares.
+		///
+		/// Parameters:
+		/// - `origin`: Must be T::CreatePoolOrigin
+		/// - `assets`: Asset ids tuple
+		/// - `initial_liquidity`: Corresponding initial liquidity of `assets`
+		/// - `amplification`: Pool amplification
+		/// - `fee`: trade fee to be applied in sell/buy trades
+		///
+		/// Emits `PoolCreated` event when successful.
+		#[pallet::weight(<T as Config>::WeightInfo::create_pool())]
+		#[transactional]
+		pub fn create_pool(
+			origin: OriginFor<T>,
+			assets: (T::AssetId, T::AssetId),
+			initial_liquidity: (Balance, Balance),
+			amplification: u16,
+			fee: Permill,
+		) -> DispatchResult {
+			let who = T::CreatePoolOrigin::ensure_origin(origin)?;
+
+			let pool_assets: PoolAssets<T::AssetId> = assets.into();
+
+			ensure!(pool_assets.is_valid(), Error::<T>::SameAssets);
+
+			ensure!(
+				T::AmplificationRange::get().contains(&amplification),
+				Error::<T>::InvalidAmplification
+			);
+
+			ensure!(T::AssetRegistry::exists(pool_assets.0), Error::<T>::AssetNotRegistered);
+			ensure!(T::AssetRegistry::exists(pool_assets.1), Error::<T>::AssetNotRegistered);
+
+			let reserves: AssetAmounts<Balance> = initial_liquidity.into();
+
+			ensure!(reserves.is_valid(), Error::<T>::InvalidInitialLiquidity);
+
+			ensure!(
+				T::Currency::free_balance(assets.0, &who) >= initial_liquidity.0,
+				Error::<T>::BalanceTooLow,
+			);
+			ensure!(
+				T::Currency::free_balance(assets.1, &who) >= initial_liquidity.1,
+				Error::<T>::BalanceTooLow,
+			);
+
+			let share_asset_ident = T::ShareAccountId::name(&pool_assets, Some(POOL_IDENTIFIER));
+
+			let share_asset = T::AssetRegistry::get_or_create_shared_asset(
+				share_asset_ident,
+				(&pool_assets).into(),
+				Balance::zero(),
```

Comment:
This is the ED parameter, no?
This would imply that I can spread my shares around as dust IIUC

Reply:
Yes,in theory you can.

The same applies to current xyk implementation. 

What would you suggest to do? Lock? 

But is it a real problem?

Reply:
Wait, it should not be zero. There should be T::MinPoolLiquidity.

Reply:
Change to correct value.

Reply:
Could it happen that someone's shares on liquidity provision are below `MinPoolLiquidity`?

Reply:
i will add a check to make sure that shares for LP are at least MinPoolLiquiduity.  I believe we have similar check in xyk implementation. 

I will also add another check in remove liquidity, to make that remaining shares are above ED.

with that , we should be ok. wdyt ?